tests/src/Test/Tasty/Vivado.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Tasty.Vivado where
+
+import Data.Coerce (coerce)
+import Data.Tagged (Tagged (..))
+import Data.Proxy (Proxy (..))
+import System.IO (hPutStr, hFlush)
+
+import System.IO.Temp (withSystemTempFile)
+import Test.Tasty.Providers (IsTest (..), testPassed)
+import Test.Tasty.Options (IsOption (..), safeReadBool, flagCLParser, lookupOption, OptionDescription (..))
+
+import Clash.Vivado (fromModuleName, HdlSource (..))
+import Clash.Annotations.Primitive (HDL (..))
+import Test.Tasty.Program
+
+-- | @--vivado@ flag for enabling tests that use vivado.
+newtype Vivado = Vivado Bool
+
+instance IsOption Vivado where
+  defaultValue = Vivado True
+  parseValue = fmap Vivado . safeReadBool
+  optionName = pure "no-vivado"
+  optionHelp = pure "Skip Vivado tests"
+  optionCLParser = flagCLParser Nothing (Vivado False)
+
+data VivadoTest = VivadoTest
+  { hdlTarget :: HDL
+  , hdlSource :: IO FilePath
+  , moduleTest :: String
+  , entity :: String
+  }
+
+instance IsTest VivadoTest where
+  run optionSet (VivadoTest target dir m top) progressCallback
+    | Vivado True <- lookupOption optionSet =
+        withSystemTempFile "sim.tcl" $ \fp h -> do
+          hdlSrc <- dir
+          tcl <- fromModuleName (HdlSource hdlSrc target) m top
+          hPutStr h tcl *>
+            hFlush h
+          runVivado hdlSrc ["-mode", "batch", "-source", fp]
+    | otherwise = pure (testPassed "Ignoring test due to --no-vivado")
+
+   where
+    vivado workDir args =
+      TestFailingProgram True "vivado" args NoGlob PrintNeither False (Just 0) (ExpectNotStdOut "expected: ") (Just workDir)
```

Comment:
I think looking for the string "expected" is too brittle. I'd prefer if we enhance our `Expect` stuff and then make the regex `^@|^Error` match. Luckily, any errors will never occur in the first line of the code, so if it is easier we could match `\n@` and `\nError`.

Also, currently Vivado outputs critical warnings, and I think those should have made CI fail:
```
CRITICAL WARNING: [Vivado 12-1504] The IP is already part of the fileset 'sources_1'. Requested source '/home/peter/clash-test-1e8af8d974db0407/xilinx_tmp/.srcs/sources_1/ip/dcfifo/dcfifo.xci' will not be added.
CRITICAL WARNING: [filemgmt 20-1440] File '/home/peter/clash-test-1e8af8d974db0407/xilinx_tmp/.gen/sources_1/ip/dcfifo/synth/dcfifo.vhd' already exists in the project as a part of sub-design file '/home/peter/clash-test-1e8af8d974db0407/xilinx_tmp/.srcs/sources_1/ip/dcfifo/dcfifo.xci'. Explicitly adding the file outside the scope of the sub-design can lead to unintended behaviors and is not recommended.
```

 I looked into it and it turns out you can actually change the severity of most of Vivado's messages, so we can elevate all critical warnings to a level where Vivado exits with an error exit code:
```
set_msg_config -severity {CRITICAL WARNING} -new_severity ERROR
```
That should be added near the top of the Tcl CI script.

PS: Our `assert` function outputs a message prefixed with the word "Error: " in VHDL and with `@` in Verilog, that's the reason for the regex.